frontend/javascripts/dashboard/dataset/sample_datasets_modal.js

```diff
@@ -0,0 +1,146 @@
+// @flow
+import { Spin, Button, Modal, List } from "antd";
+import React, { useState, useEffect } from "react";
+import _ from "lodash";
+
+import {
+  getSampleDatasets,
+  triggerSampleDatasetDownload,
+  getDatastores,
+} from "admin/admin_rest_api";
+import { useInterval } from "libs/react_helpers";
+import { handleGenericError } from "libs/error_handling";
+import type { APIDataStore, APISampleDataset } from "admin/api_flow_types";
+
+type Props = {
+  destroy: () => void,
+  onOk?: () => any,
+  organizationName: string,
+};
+
+function useDatastores(): [Array<APIDataStore>] {
+  const [datastores, setDatastores] = useState([]);
+
+  const fetchDatastores = async () => {
+    const fetchedDatastores = await getDatastores();
+    setDatastores(fetchedDatastores);
+  };
+
+  useEffect(() => {
+    fetchDatastores();
+  }, []);
+  return [datastores];
+}
+
+function useSampleDatasets(
+  datastore,
+  organizationName,
+  pendingDatasets,
+  setPendingDatasets,
+): [Array<APISampleDataset>, Array<string>, () => Promise<void>] {
+  const [datasets, setDatasets] = useState([]);
+  const [failedDatasets, setFailedDatasets] = useState([]);
+
+  const fetchSampleDatasets = async () => {
+    if (datastore == null) return;
+    const sampleDatasets = await getSampleDatasets(datastore.url, organizationName);
+    setDatasets(sampleDatasets);
+
+    // Datasets that were pending, but are now available again, failed to download
+    setFailedDatasets(
+      failedDatasets.concat(
+        pendingDatasets.filter(pendingDataset =>
+          sampleDatasets.find(
+            dataset => dataset.name === pendingDataset && dataset.status === "available",
+          ),
+        ),
+      ),
+    );
+
+    // Remove datasets which are not downloading from the pendingDatasets queue
+    setPendingDatasets(
+      _.without(
+        pendingDatasets,
+        ...sampleDatasets
+          .filter(dataset => dataset.status !== "downloading")
+          .map(dataset => dataset.name),
+      ),
+    );
+  };
+
+  useEffect(() => {
+    fetchSampleDatasets();
+  }, [datastore]);
+
+  return [datasets, failedDatasets, fetchSampleDatasets];
+}
+
+const SampleDatasetsModal = ({ destroy, onOk, organizationName }: Props) => {
+  const [pendingDatasets, setPendingDatasets] = useState([]);
+  const [datastores] = useDatastores();
+  const datastore = datastores[0];
+  const [datasets, failedDatasets, fetchDatasets] = useSampleDatasets(
+    datastore,
+    organizationName,
+    pendingDatasets,
+    setPendingDatasets,
+  );
+
+  useInterval(fetchDatasets, pendingDatasets.length ? 1000 : null);
+
+  const handleSampleDatasetDownload = async (name: string) => {
```

Comment:
I'm wondering whether it makes sense to move `handleSampleDatasetDownload`, the `pendingDatasets` definition and the `useDatastores` call into the `useSampleDatasets` hook. At the moment, it feels a bit weird that `failedDatasets` are provided by the `useSampleDatasets` hook, but the `pendingDatasets` have to be maintained at an higher level.
`handleSampleDatasetDownload` would be another return argument of `useSampleDatasets`.

This would clean up the view logic here a bit, but obviously the `useSampleDatasets` hook would get longer. I'd argue, that the abstraction level is more "even", though. What do you think?

Reply:
I struggled with that as well, thinking along the same lines ^^ I'll give this option a try and see what I can do to clean this up a little bit :)